include/range/v3/view/cartesian_product.hpp

```diff
@@ -435,11 +439,38 @@ namespace ranges
                 return cartesian_product_view<all_t<Rngs>...>{
                     all(static_cast<Rngs &&>(rngs))...};
             }
-
-            constexpr empty_view<std::tuple<>> operator()() const noexcept
+#if defined(_MSC_VER)
+            template<typename Rng0>
+            constexpr auto operator()(Rng0 && rng0) const
+                -> CPP_ret(cartesian_product_view<all_t<Rng0>>)( //
+                    requires forward_range<Rng0> && viewable_range<Rng0>)
             {
-                return {};
+                return cartesian_product_view<all_t<Rng0>>{
+                    all(static_cast<Rng0 &&>(rng0))};
+            }
+            template<typename Rng0, typename Rng1>
+            constexpr auto operator()(Rng0 && rng0, Rng1 && rng1) const
+                -> CPP_ret(cartesian_product_view<all_t<Rng0>, all_t<Rng1>>)( //
+                    requires forward_range<Rng0> && viewable_range<Rng0> &&   //
+                             forward_range<Rng1> && viewable_range<Rng1>)
+            {
+                return cartesian_product_view<all_t<Rng0>, all_t<Rng1>>{
+                    all(static_cast<Rng0 &&>(rng0)), //
+                    all(static_cast<Rng1 &&>(rng1))};
             }
+            template<typename Rng0, typename Rng1, typename Rng2>
+            constexpr auto operator()(Rng0 && rng0, Rng1 && rng1, Rng2 && rng2) const
+                -> CPP_ret(cartesian_product_view<all_t<Rng0>, all_t<Rng1>, all_t<Rng2>>)( //
+                    requires forward_range<Rng0> && viewable_range<Rng0> &&            //
+                             forward_range<Rng1> && viewable_range<Rng1> &&
+                             forward_range<Rng2> && viewable_range<Rng2>)
+            {
+                return cartesian_product_view<all_t<Rng0>, all_t<Rng1>, all_t<Rng2>>{
+                    all(static_cast<Rng0 &&>(rng0)), //
+                    all(static_cast<Rng1 &&>(rng1)), //
+                    all(static_cast<Rng2 &&>(rng2))};
+            }
+#endif
```

Comment:
And this is an example of where I had to avoid variadic arguments in `operator()`.